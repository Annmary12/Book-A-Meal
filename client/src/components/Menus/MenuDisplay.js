import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';
import NoMenu from '../NoMenu/NoMenu.js';
import MenuCategory from './MenuCategory.js';
import MenuNavigator from './MenuNavigator.js';
import styles from './menus.scss';

/**
 * @class MenuDisplay
 */
class MenuDisplay extends Component {
  static propTypes = {
    menus: PropTypes.arrayOf(PropTypes.object).isRequired,
    loggedIn: PropTypes.bool.isRequired
  }

  state = {
    currentMenuIndex: 0
  }

  goToNextMenu = () => {
    const { currentMenuIndex } = this.state;

    if (this.nextExists()) {
      this.flipMenuBody();
      this.setState({ currentMenuIndex: currentMenuIndex + 1 });
    }
  }

  goToPrevMenu = () => {
    const { currentMenuIndex } = this.state;

    if (this.prevExists()) {
      this.flipMenuBody();

      this.setState({ currentMenuIndex: currentMenuIndex - 1 });
    }
  }

  prevExists = () => {
    const { currentMenuIndex } = this.state;

    return currentMenuIndex > 0;
  }

  nextExists = () => {
    const { currentMenuIndex } = this.state;
    const { menus } = this.props;

    return currentMenuIndex < menus.length - 1;
  }

  /**
   * @return {undefined} undefined
   */
  flipMenuBody() {
    this.menuBody.classList.toggle(styles.flipIn);
    setTimeout(() => this.menuBody.classList.toggle(styles.flipIn), 400);
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    const { currentMenuIndex } = this.state;
    const { menus, loggedIn } = this.props;

    return isEmpty(menus) ?
      <div className={styles.noMenu}>
        <NoMenu loggedIn={loggedIn} />
      </div> :
      <div className={styles.slideInUp}>
        <MenuNavigator
          prevExists={this.prevExists}
          nextExists={this.nextExists}
          goToNextMenu={this.goToNextMenu}
          goToPrevMenu={this.goToPrevMenu}
        />
        <div ref={el => this.menuBody = el} className={styles.menuBody}>
          <div className={styles.catererName}>
            {menus[currentMenuIndex].caterer.username}&apos;s menu
          </div>
          {
            menus[currentMenuIndex].categories.map(category => (
              <MenuCategory key={category.id} category={category} />
            ))
          }
        </div>
      </div>;
  }
}

export default MenuDisplay;
