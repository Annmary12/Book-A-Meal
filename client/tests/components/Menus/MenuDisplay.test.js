import React from 'react';
import MenuDisplay from '../../../src/components/Menus/MenuDisplay';

const menus = [{
  "id": "d48a03fa-ca24-4f5c-9ff2-61c6f5b7b633",
  "date": "2018-06-24",
  "categories": [{
    "id": "a7166d2c-3af2-438e-ba70-aa02d1a72fb5",
    "title": "Appetizer",
    "meals": [{
      "id": "e0a8e7f2-c27b-484e-9cd7-c4720e14b004",
      "title": "Rice and stew",
      "description": "Delicious Nigerian rice and Chicken Stew with Onions, Tomatoes, and Dijon Recipe",
      "price": 1200,
      "img": "https://www.simplyrecipes.com/wp-content/uploads/2009/04/chicken-stew-onions-tomatoes-dijon-horiz-a-1600.jpg"
    }]
  }],
  "caterer": {
    "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
    "fullName": "Otse Obande",
    "username": "otseobande",
    "email": "mealbooker@gmail.com",
    "role": "caterer"
  }
}];


describe('MenuDisplay component', () => {
  it('should match snapshot if menus array is empty', () => {
    const wrapper = shallow(
      <MenuDisplay
        menus={[]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if menus array is not empty', () => {
    const wrapper = shallow(
      <MenuDisplay
        menus={menus}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })
});