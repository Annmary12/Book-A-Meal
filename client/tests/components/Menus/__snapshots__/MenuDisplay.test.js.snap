// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuDisplay component should match snapshot if menus array is empty 1`] = `
<div
  className="noMenu"
>
  <NoMenu />
</div>
`;

exports[`MenuDisplay component should match snapshot if menus array is not empty 1`] = `
<div
  className="slideInUp"
>
  <MenuNavigator
    goToNextMenu={[Function]}
    goToPrevMenu={[Function]}
    nextExists={[Function]}
    prevExists={[Function]}
  />
  <div
    className="menuBody"
  >
    <div
      className="catererName"
    >
      otseobande
      's menu
    </div>
    <MenuCategory
      category={
        Object {
          "id": "a7166d2c-3af2-438e-ba70-aa02d1a72fb5",
          "meals": Array [
            Object {
              "description": "Delicious Nigerian rice and Chicken Stew with Onions, Tomatoes, and Dijon Recipe",
              "id": "e0a8e7f2-c27b-484e-9cd7-c4720e14b004",
              "img": "https://www.simplyrecipes.com/wp-content/uploads/2009/04/chicken-stew-onions-tomatoes-dijon-horiz-a-1600.jpg",
              "price": 1200,
              "title": "Rice and stew",
            },
          ],
          "title": "Appetizer",
        }
      }
      key="a7166d2c-3af2-438e-ba70-aa02d1a72fb5"
    />
  </div>
</div>
`;
